lib/zendesk_apps_support/validations/translations.rb

```diff
@@ -48,12 +48,56 @@ def json_error(file)
           ValidationError.new('translation.not_json', file: file.relative_path, errors: e)
         end
 
-        def required_keys(file)
-          return if file.relative_path != 'translations/en.json' || JSON.parse(file.read)['app']['name']
+        def validate_marketplace_content(file, package)
+          return [] if file.relative_path != 'translations/en.json'
+
+          errors = []
+          json = JSON.parse(file.read)
+          product_names = Product::PRODUCTS_AVAILABLE.map(&:name)
+          present_product_keys = json['app'].keys & product_names
+
+          if present_product_keys.empty?
+            errors << validate_top_level_required_keys(json, file.relative_path)
+          else
+            errors << validate_products_match_manifest_products(present_product_keys, package, file.relative_path)
+            errors << validate_products_have_required_keys(json, present_product_keys, file.relative_path)
+          end
+          errors.compact
+        end
+
+        def validate_top_level_required_keys(json, file_path)
+          return if (json['app'].keys & MANDATORY_KEYS).sort == MANDATORY_KEYS.sort
+          missing_keys = MANDATORY_KEYS - json['app'].keys
+          ValidationError.new(
+            'translation.missing_required_key',
+            file: file_path,
+            missing_key: missing_keys.join(', ')
+          )
+        end
+
+        def validate_products_have_required_keys(json, products, file_path)
+          products.each do |product|
+            next unless (json['app'][product].keys & MANDATORY_KEYS).sort != MANDATORY_KEYS.sort
```

Comment:
After the refactor I don't think this looks like as good an idea - it's just an array subtraction. I'd rather leave it as is.